File: TicTacToe.Server/Models/Game.cs
namespace TicTacToe.Server.Models;

public class Game
{
    public string Id { get; set; } = Guid.NewGuid().ToString();
    public Player? Player1 { get; set; }
    public Player? Player2 { get; set; }
    public int BoardSize { get; set; } = 3;
    public CellState[,] Board { get; set; }
    public GameState State { get; set; } = GameState.WaitingForPlayers;
    public GameResult Result { get; set; } = GameResult.None;
    public Player? CurrentPlayer { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? StartedAt { get; set; }
    public DateTime? EndedAt { get; set; }
    public int MoveCount { get; set; } = 0;
    public bool IsPrivate { get; set; } = false;

    public Game(int boardSize = 3)
    {
        BoardSize = boardSize;
        Board = new CellState[boardSize, boardSize];
    }

    public bool IsPlayerTurn(string playerId)
    {
        return CurrentPlayer?.Id == playerId;
    }

    public bool IsGameFull()
    {
        return Player1 != null && Player2 != null;
    }

    public bool IsValidMove(int row, int col)
    {
        return row >= 0 && row < BoardSize && col >= 0 && col < BoardSize && Board[row, col] == CellState.Empty;
    }

    private CellState CharacterIconToCellState(CharacterIcon icon)
    {
        return icon switch
        {
            CharacterIcon.Cross => CellState.Cross,
            CharacterIcon.Circle => CellState.Circle,
            CharacterIcon.Kuromi => CellState.Kuromi,
            CharacterIcon.MyMelody => CellState.MyMelody,
            CharacterIcon.Spiderman => CellState.Spiderman,
            CharacterIcon.Cinnamoroll => CellState.Cinnamoroll,
            CharacterIcon.BadtzMaru => CellState.BadtzMaru,
            CharacterIcon.HelloKitty => CellState.HelloKitty,
            CharacterIcon.Keroppi => CellState.Keroppi,
            CharacterIcon.Pochacco => CellState.Pochacco,
            CharacterIcon.AI => CellState.AI,
            _ => CellState.Cross
        };
    }

    public void MakeMove(int row, int col, CharacterIcon characterIcon)
    {
        if (IsValidMove(row, col))
        {
            Board[row, col] = CharacterIconToCellState(characterIcon);
            MoveCount++;
            
            // Switch turns
            CurrentPlayer = CurrentPlayer?.Id == Player1?.Id ? Player2 : Player1;
        }
    }

    private CellState GetPlayerCellState(Player player)
    {
        return CharacterIconToCellState(player.CharacterIcon);
    }

    public GameResult CheckWinner()
    {
        if (Player1 == null || Player2 == null) return GameResult.None;

        var player1CellState = GetPlayerCellState(Player1);
        var player2CellState = GetPlayerCellState(Player2);
        var winLength = BoardSize; // For dynamic win condition: 3 in a row for 3x3, 4 in a row for 4x4

        // Check rows
        for (int row = 0; row < BoardSize; row++)
        {
            if (CheckLine(row, 0, 0, 1, winLength, player1CellState, player2CellState) != GameResult.None)
                return CheckLine(row, 0, 0, 1, winLength, player1CellState, player2CellState);
        }

        // Check columns
        for (int col = 0; col < BoardSize; col++)
        {
            if (CheckLine(0, col, 1, 0, winLength, player1CellState, player2CellState) != GameResult.None)
                return CheckLine(0, col, 1, 0, winLength, player1CellState, player2CellState);
        }

        // Check main diagonal (top-left to bottom-right)
        if (CheckLine(0, 0, 1, 1, winLength, player1CellState, player2CellState) != GameResult.None)
            return CheckLine(0, 0, 1, 1, winLength, player1CellState, player2CellState);

        // Check anti-diagonal (top-right to bottom-left)
        if (CheckLine(0, BoardSize - 1, 1, -1, winLength, player1CellState, player2CellState) != GameResult.None)
            return CheckLine(0, BoardSize - 1, 1, -1, winLength, player1CellState, player2CellState);

        // Check for draw (all cells filled)
        if (MoveCount >= BoardSize * BoardSize)
        {
            return GameResult.Draw;
        }

        return GameResult.None;
    }

    private GameResult CheckLine(int startRow, int startCol, int deltaRow, int deltaCol, int winLength, CellState player1CellState, CellState player2CellState)
    {
        CellState firstCell = Board[startRow, startCol];
        if (firstCell == CellState.Empty) return GameResult.None;

        for (int i = 1; i < winLength; i++)
        {
            int row = startRow + i * deltaRow;
            int col = startCol + i * deltaCol;
            
            if (row >= BoardSize || col >= BoardSize || col < 0 || Board[row, col] != firstCell)
                return GameResult.None;
        }

        return firstCell == player1CellState ? GameResult.Player1Wins : GameResult.Player2Wins;
    }
} 

File: tictactoe.client/src/types/game.ts
export const CharacterIcon = {
  Cross: 'Cross',
  Circle: 'Circle',
  Kuromi: 'Kuromi',
  MyMelody: 'MyMelody',
  Spiderman: 'Spiderman',
  Cinnamoroll: 'Cinnamoroll',
  BadtzMaru: 'BadtzMaru',
  HelloKitty: 'HelloKitty',
  Keroppi: 'Keroppi',
  Pochacco: 'Pochacco',
  AI: 'AI'
} as const;

export type CharacterIcon = typeof CharacterIcon[keyof typeof CharacterIcon];

export interface Player {
  id: string;
  name: string;
  characterIcon: CharacterIcon;
  isReady: boolean;
  isActive: boolean;
  joinedAt: Date;
}

export interface Game {
  id: string;
  player1: Player | null;
  player2: Player | null;
  board: string[][];
  state: GameState;
  result: GameResult;
  currentPlayer: Player | null;
  createdAt: Date;
  startedAt: Date | null;
  endedAt: Date | null;
  moveCount: number;
  isPrivate: boolean;
}

export const GameState = {
  WaitingForPlayers: 'WaitingForPlayers',
  InProgress: 'InProgress',
  Finished: 'Finished'
} as const;

export type GameState = typeof GameState[keyof typeof GameState];

export const GameResult = {
  None: 'None',
  Player1Wins: 'Player1Wins',
  Player2Wins: 'Player2Wins',
  Draw: 'Draw'
} as const;

export type GameResult = typeof GameResult[keyof typeof GameResult];

export interface GameListItem {
  gameId: string;
  player1Name: string;
  player1CharacterIcon: CharacterIcon;
  createdAt: Date;
  isPrivate: boolean;
}

export interface GameCreatedEvent {
  gameId: string;
  playerId: string;
  playerName: string;
  characterIcon: CharacterIcon;
  state: string;
  isPrivate: boolean;
}

export interface GameStartedEvent {
  gameId: string;
  player1: Player;
  player2: Player;
  currentPlayer: Player;
  state: string;
}

export interface MoveMadeEvent {
  gameId: string;
  row: number;
  col: number;
  playerId: string;
  board: string[][];
  currentPlayer: Player | null;
  moveCount: number;
}

export interface GameFinishedEvent {
  gameId: string;
  result: string;
  winner: string | null;
  endedAt: Date;
}

export interface GameResetEvent {
  gameId: string;
  board: string[][];
  currentPlayer: Player;
  state: string;
}

export interface PlayerDisconnectedEvent {
  gameId: string;
  playerId: string;
  playerName: string;
}

export interface GameError {
  message: string;
}

export interface CharacterIconInfo {
  icon: CharacterIcon;
  displayName: string;
  fileName: string;
}

export const CHARACTER_ICONS: CharacterIconInfo[] = [
  { icon: CharacterIcon.Cross, displayName: 'Cross', fileName: 'icons8-cross-100.png' },
  { icon: CharacterIcon.Circle, displayName: 'Circle', fileName: 'icons8-circle-100-4.png' },
  { icon: CharacterIcon.Kuromi, displayName: 'Kuromi', fileName: 'icons8-kuromi-100.png' },
  { icon: CharacterIcon.MyMelody, displayName: 'My Melody', fileName: 'icons8-my-melody-100.png' },
  { icon: CharacterIcon.Spiderman, displayName: 'Spider-Man', fileName: 'icons8-spiderman-100-2.png' },
  { icon: CharacterIcon.Cinnamoroll, displayName: 'Cinnamoroll', fileName: 'icons8-cinnamoroll-100.png' },
  { icon: CharacterIcon.BadtzMaru, displayName: 'Badtz-Maru', fileName: 'icons8-badtz-maru-100.png' },
  { icon: CharacterIcon.HelloKitty, displayName: 'Hello Kitty', fileName: 'icons8-hello-kitty-100.png' },
  { icon: CharacterIcon.Keroppi, displayName: 'Keroppi', fileName: 'icons8-keroppi-100.png' },
  { icon: CharacterIcon.Pochacco, displayName: 'Pochacco', fileName: 'icons8-pochacco-100.png' },
  { icon: CharacterIcon.AI, displayName: 'AI', fileName: 'icons8-ai-100.png' }
];

// Re-export GameDto types for convenience
export type { GameDto, BaseDto, CreateGameRequest, UpdateGameRequest } from './gameDto'; 